flow_action_components/GetOpportunityPartners/force-app/main/default/classes/GetOpportunityPartnersAction.cls

```diff
@@ -0,0 +1,132 @@
+/**
+* @author       : Tamar Erlich
+* @date         : October 07, 2020
+* @description  : Invocable method that given opportunityId and accountId, returns a list of opportunity partners to display on opportunity plan record page
+* Note          : This class is called by the Flow.
+* *************************************************
+*       <runTest>GetOpportunityPartnersActionTest<runTest>
+* *************************************************
+* @History
+* -------
+* VERSION | AUTHOR                | DATE               | DESCRIPTION
+* 1.0     | Tamar Erlich          | October 07, 2020   | Created
+**/
+global with sharing class GetOpportunityPartnersAction {
+
+    // Attributes passed in from the Flow
+    global class Requests {
+
+        @InvocableVariable(label='OpportunityId' required=true)
+        global String opportunityId;
+        
+        @InvocableVariable(label='AccountId' required=true)
+        global String accountId;
+    }
+
+    // Attributes passed back to the Flow
+    global class Results {
+
+        public Results() {
+            
+        }
+        @InvocableVariable
+        global String errors;
+
+        @InvocableVariable
+        global String successful;
+
+        @InvocableVariable
+        public List<OpportunityPartnersWrapper> outputCollection;
+
+        @InvocableVariable
+        public String outputString;
+        
+    }
+
+    // custom exception class
+    global class InvocableActionException extends Exception {}
+
+    // Expose this Action to the Flow
+    @InvocableMethod
+        global static List <Results> get(List<Requests> requestList) {
+            // initialize variables
+            Results response = new Results();
+            List<Results> responseWrapper = new List<Results>();
+            String errors;
+            String success = 'true';
+            String stringOutput;
+            String accountId = requestList[0].accountId;
```

Comment:
Most actions created through 2019 used this [0] approach, but it was pointed out that they lost the ability to work in batch scenarios. So these days, we instead use the for loop approach that you can see here: https://github.com/alexed1/LightningFlowComponents/blob/master/flow_action_components/CollectionProcessors/force-app/main/default/classes/CopyCollection.cls

Once you do it once or twice it becomes very natural.

Reply:
I tried to add the for loop but then I got a SOQL query inside a loop error

Reply:
Can you show me the code where you got the error? We really don't want to publish new actions that aren't bulkified....

Reply:
Hi @alexed1 , I think I was able to bulkify the action code (and learned how to use maps along the way :) ) but I would like to understand why it is relevant here since we are ever only passing a single opportunityId into the action?

Reply:
Most bulkification scenarios start with 'suppose someone uploads a batch of 50 opportunities, and that triggers a flow that has this action?'  Flow will try to send all 50 to this action. It's good for performance.

Reply:
Got your point, thanks for explaining.
Is there anything else I need to refactor in this action?